File: Harvest/Services/ProfileService.swift
import Foundation
import Supabase

struct ProfileService {
    private var client: SupabaseClient { SupabaseManager.shared.client }

    func getProfile(userId: String) async throws -> UserProfile? {
        let response: [UserProfile] = try await client
            .from("users")
            .select()
            .eq("id", value: userId)
            .execute()
            .value

        guard let profile = response.first else { return nil }

        return profile
    }

    func createProfile(userId: String, email: String) async throws -> UserProfile? {
        let defaultNickname = email.components(separatedBy: "@").first ?? "User"
        let now = ISO8601DateFormatter().string(from: Date())

        let values: [String: String] = [
            "id": userId,
            "email": email,
            "nickname": defaultNickname,
            "bio": "I'm new here!",
            "created_at": now,
            "updated_at": now
        ]

        let response: [UserProfile] = try await client
            .from("users")
            .upsert(values, onConflict: "id")
            .select()
            .execute()
            .value
        return response.first
    }

    func upsertProfile(userId: String, updates: [String: AnyJSON]) async throws -> UserProfile? {
        let sessionEmail = try? await client.auth.session.user.email

        let response: [UserProfile] = try await client
            .from("users")
            .upsert(
                Self.upsertPayload(
                    userId: userId,
                    email: sessionEmail,
                    updates: updates,
                    nowISO: ISO8601DateFormatter().string(from: Date())
                ),
                onConflict: "id"
            )
            .select()
            .execute()
            .value
        return response.first
    }

    /// The row an upsert proposes.
    ///
    /// `users.email` is NOT NULL with no default, and Postgres validates the
    /// proposed tuple *before* it resolves ON CONFLICT — so leaving email out
    /// raises 23502 whether or not the row already exists, which is what
    /// stranded users at the last step of onboarding.
    ///
    /// The session's address goes in first so an explicit one in `updates`
    /// still wins, and an unknown one is left out rather than written as null.
    static func upsertPayload(
        userId: String,
        email: String?,
        updates: [String: AnyJSON],
        nowISO: String
    ) -> [String: AnyJSON] {
        var payload: [String: AnyJSON] = [:]
        if let email { payload["email"] = .string(email) }
        payload.merge(updates) { _, new in new }
        payload["id"] = .string(userId)
        payload["updated_at"] = .string(nowISO)
        return payload
    }

    func updateProfile(userId: String, updates: [String: AnyJSON]) async throws -> UserProfile? {
        var mutableUpdates = updates
        mutableUpdates["updated_at"] = AnyJSON.string(ISO8601DateFormatter().string(from: Date()))

        let response: [UserProfile] = try await client
            .from("users")
            .update(mutableUpdates)
            .eq("id", value: userId)
            .select()
            .execute()
            .value
        return response.first
    }

    func checkOnboardingStatus(userId: String) async throws -> Bool {
        struct OnboardingCheck: Decodable {
            let bio: String?
            let age: Int?
            let gender: String?
            let photos: [String]?
        }

        let response: [OnboardingCheck] = try await client
            .from("users")
            .select("bio, age, gender, photos")
            .eq("id", value: userId)
            .execute()
            .value

        guard let data = response.first else { return false }
        return data.bio != nil && data.age != nil && data.gender != nil && (data.photos?.isEmpty == false)
    }

    func uploadPhoto(userId: String, imageData: Data, photoIndex: Int) async throws -> String {
        let fileName = "\(userId)/photo_\(photoIndex)_\(Int(Date().timeIntervalSince1970 * 1000)).jpg"

        let response = try await client.storage
            .from(Config.storageBucket)
            .upload(
                fileName,
                data: imageData,
                options: FileOptions(contentType: "image/jpeg", upsert: true)
            )

        return "\(Config.supabaseURL)/storage/v1/object/public/\(response.fullPath)"
    }

    func deletePhoto(userId: String, photoUrl: String) async throws {
        guard let url = URL(string: photoUrl),
              let pathComponent = url.path.components(separatedBy: "/storage/v1/object/public/\(Config.storageBucket)/").last
        else { return }

        try await client.storage
            .from(Config.storageBucket)
            .remove(paths: [pathComponent])
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/data/service/ProfileService.kt
package com.harvestglass.harvest.data.service

import com.harvestglass.harvest.Config
import com.harvestglass.harvest.data.model.UserProfile
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.storage.storage
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.time.Instant

/** Mirrors Harvest/Services/ProfileService.swift. */
class ProfileService(private val client: SupabaseClient) {

    suspend fun getProfile(userId: String): UserProfile? =
        client.postgrest.from("users")
            .select { filter { eq("id", userId) } }
            .decodeList<UserProfile>()
            .firstOrNull()

    suspend fun createProfile(userId: String, email: String): UserProfile? {
        val now = Instant.now().toString()
        return client.postgrest.from("users")
            .upsert(
                buildJsonObject {
                    put("id", userId)
                    put("email", email)
                    put("nickname", defaultNickname(email))
                    put("bio", "I'm new here!")
                    put("created_at", now)
                    put("updated_at", now)
                }
            ) {
                onConflict = "id"
                select()
            }
            .decodeList<UserProfile>()
            .firstOrNull()
    }

    suspend fun upsertProfile(userId: String, updates: JsonObject): UserProfile? =
        client.postgrest.from("users")
            .upsert(
                buildJsonObject {
                    updates.forEach { (k, v) -> put(k, v) }
                    put("id", userId)
                    put("updated_at", Instant.now().toString())
                }
            ) {
                onConflict = "id"
                select()
            }
            .decodeList<UserProfile>()
            .firstOrNull()

    suspend fun updateProfile(userId: String, updates: JsonObject): UserProfile? =
        client.postgrest.from("users")
            .update(
                buildJsonObject {
                    updates.forEach { (k, v) -> put(k, v) }
                    put("updated_at", Instant.now().toString())
                }
            ) {
                filter { eq("id", userId) }
                select()
            }
            .decodeList<UserProfile>()
            .firstOrNull()

    suspend fun uploadPhoto(userId: String, imageData: ByteArray, photoIndex: Int): String {
        val path = photoObjectPath(userId, photoIndex, System.currentTimeMillis())
        val result = client.storage.from(Config.STORAGE_BUCKET).upload(path, imageData) {
            upsert = true
        }
        return "${Config.SUPABASE_URL}/storage/v1/object/public/${result.path}"
    }

    suspend fun deletePhoto(userId: String, photoUrl: String) {
        // Swift returns silently on an unparseable URL rather than throwing;
        // a malformed row shouldn't block the user from removing the photo.
        val path = storagePathFromUrl(photoUrl) ?: return
        client.storage.from(Config.STORAGE_BUCKET).delete(path)
    }

    // MARK: - Pure helpers, unit-tested without a network

    internal fun photoObjectPath(userId: String, photoIndex: Int, epochMillis: Long): String =
        "$userId/photo_${photoIndex}_$epochMillis.jpg"

    /** The object path inside our bucket, or null when the URL isn't one of ours. */
    internal fun storagePathFromUrl(photoUrl: String): String? {
        val marker = "/storage/v1/object/public/${Config.STORAGE_BUCKET}/"
        val index = photoUrl.indexOf(marker)
        if (index < 0) return null
        return photoUrl.substring(index + marker.length).ifEmpty { null }
    }

    internal fun defaultNickname(email: String): String =
        email.substringBefore('@').ifEmpty { "User" }
}
